### DISCUSSION TITLE
AppRouter import into createTRPCNext() trpc client throwing type error

### DISCUSSION THREAD

**Discussion by Eduard-Hasa** [trpc/trpc#3102] [Q&A] [answered]
Im trying to use trpc server on a Cloudflare Worker on a trpc client on next js but getting this strange type error. Does anyone know what might be going on here and how I could fix it? Thanks in advance.

<img width="1382" alt="image" src="https://user-images.githubusercontent.com/48266101/199387544-2ad0f277-1fce-4069-a7a6-de97fbd31c42.png">

<img width="946" alt="image" src="https://user-images.githubusercontent.com/48266101/199387651-25266c35-21d3-422d-b930-051f71e42f26.png">

**Comment by JonParton** [comment]
Likely its because your [`input` has to be an object](https://trpc.io/docs/v10/procedures#input-validation) .

So

```ts
hello: publicProcedure.input(z.object({name : z.string().nullish()}).query(({input}) => {.....}) 
```


P. S. Sorry for the formatting, I'm on mobile!

**Reply by Eduard-Hasa** [reply]
Hmm unfortunately that does not seem to be the cause. I copied my hello query from official trpc repos but I even tried yours and even no query at all and still getting errors. 


Type 'Router<RouterDef<RootConfig<{ ctx: {}; meta: {}; errorShape: DefaultErrorShape; transformer: DefaultDataTransformer; }>, {}, { queries: {}; mutations: {}; subscriptions: {}; }>>' does not satisfy the constraint 'Router<AnyRouterDef<AnyRootConfig, any>>'.
  The types returned by 'createCaller(...)' are incompatible between these types.
    Type '{ query: inferHandlerFn<{}>; mutation: inferHandlerFn<{}>; subscription: inferHandlerFn<{}>; } & DecoratedProcedureRecord<{}>' is not assignable to type '{ query: inferHandlerFn<any>; mutation: inferHandlerFn<any>; subscription: inferHandlerFn<any>; } & DecoratedProcedureRecord<any>'.
      Type '{ query: inferHandlerFn<{}>; mutation: inferHandlerFn<{}>; subscription: inferHandlerFn<{}>; } & DecoratedProcedureRecord<{}>' is not assignable to type '{ query: inferHandlerFn<any>; mutation: inferHandlerFn<any>; subscription: inferHandlerFn<any>; }'.
        Types of property 'query' are incompatible.
          Type 'inferHandlerFn<{}>' is not assignable to type 'inferHandlerFn<any>'.
            Types of parameters 'path' and 'path' are incompatible.
              Type 'TPath' is not assignable to type 'never'.
                Type 'string' is not assignable to type 'never'.ts(2344)

**Reply by JonParton** [reply]
Hmmmn....

Just from your screen shots I can't see exactly what is wrong but I did notice on the client side you are setting it up with a superjson transformer and on the server side you are not...

Maybe take a look at how it is initialised in the starter project with the context and transformer etc and see if that fixes it?

https://github.com/trpc/examples-next-prisma-starter/blob/next/src/server/trpc.ts

(still on mobile and just working from your screenshots so might not be exactly on point but the example project shows you a completely working example!)

**Reply by Eduard-Hasa** [reply]
Hmm I guess let me ask you this. Am I able to use trpc in nextjs as a client and cloudflare workers on the server side? I am able to get it working using next js api acting as the server but trying to take advantage of cloudflare workers durable objects. What I was doing is using the trpc example in the documentation for the Fetch Api on the server side and then for the client side I was just using the code generated from create-t3-app. I noticed on the client side we are using "createTRPCNext"... Do you think that is not compatible with the Fetch Api trpc server?

Thank you for your patience. I am coming from rest and discovered this blessing called trpc and just trying to figure it out lol.

**Reply by JonParton** [reply]
Hmmmn.. Looking at this repo it seems like you may have to use a fetch adapter??

https://github.com/esamattis/trpc-cloudflare-worker

However, I have not used cloudflare workers before so unfortunately I'm not the expert at all!!!

**Reply by JonParton** [reply]
Although I know we removed our hard dependence on fetch in V10 (... Possibly before)... So this might not be as big a thing as in that repo as it is 10 months old...

**Reply by JonParton** [reply]
And as its TS complaining when passing in the `appRouter ` type not an error at runtime it seems at this inital point to be more of a setup code problem (like the mismatch in transformers) rather then a fetch adapter needing to be patched...

**Reply by Eduard-Hasa** [reply]
@JonParton  I have literally tried everything and still spinning in circles with no resolution. I have tried adding a transformer.. Heck I have even tried completely copying the code that works in nextjs api server into the cloudflare worker with no luck at all.

**Reply by JonParton** [reply]
Just found this documentation

https://trpc.io/docs/v10/fetch

Is any of that usefull in helping to resolve?

Otherwise can you post a repro and I / someone could take a look what is going on?

**Reply by JonParton** [reply]
It might however be a miss match of the next implementation expecting the server to be on a next api back end too... In which case you may have to use the react implementation rather then next?... Would have to dig deeper on that one!

**Reply by Eduard-Hasa** [reply]
Yeah that's exactly what I'm thinking. It must be a mismatch between the fetch and the next router. I'll try a fresh react project and see if that works with my backend. I wonder if anyone can speak on this?

Is it possible to use the next trpc client with SRR and the cloudflare workers fetch trpc server?

**Reply by Eduard-Hasa** [reply]
> Just found this documentation
> 
> 
> 
> https://trpc.io/docs/v10/fetch
> 
> 
> 
> Is any of that usefull in helping to resolve?
> 
> 
> 
> Otherwise can you post a repro and I / someone could take a look what is going on? 

Yeah I have been reading this over and over and using their exact code with no luck. Even dug into the cloudflare workers examples in the trpc GitHub repo

**Reply by Eduard-Hasa** [reply]
@JonParton  Im officially blocked. I have literally tried every example but still getting the type error when importing AppRouter in the nextjs client. By the way trpc does work but im just really trying to figure out why i'm getting a type error.

**Comment by Eduard-Hasa** [comment, marked as answer]
Solution: I was using different versions of trpc on the server and the client. Updated them both to the same latest version and everything worked great. Thanks @JonParton for the effort.

**Reply by JonParton** [reply]
Fantastic!

It's always the annoying little things that trip you up and are rediculous to find.. but I'm glad you got there!! 🙂🎉

Enjoy the type safety and hopefully the productivity now you have it sorted!!!
